src/wallet/rpcwallet.cpp

```diff
@@ -2397,6 +2399,42 @@ UniValue lockunspent(const JSONRPCRequest& request)
 
         COutPoint outpt(uint256S(txid), nOutput);
 
+        const bool is_locked = pwallet->IsLockedCoin(outpt.hash, outpt.n);
+
+        if (fUnlock && !is_locked) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, expected locked output");
+        }
+
+        if (!fUnlock) {
+            if (is_locked) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, output already locked");
+            }
+
+            if (pwallet->IsSpent(outpt.hash, outpt.n)) {
```

Comment:
Why are these three checks ("expected unspent", "unknown transaction" and "vout index out of bounds") only done when trying to lock? Isn't it equally invalid to try to unlock bad txouts? Can you place these three checks above the `const bool is_locked = pwallet->IsLockedCoin(outpt.hash, outpt.n);` line?

Reply:
The idea was that in all three of those cases (already spent output, unknown transaction, and vout index out of bounds), `IsLockedCoin` should already have returned false, so checking `!is_locked` should be enough.

Reply:
I agree that the function will correctly return an error, but I think that it's better to return the more precise error if possible. `expected unspent output`, `unknown transaction` and `vout index out of bounds` are better to return than `expected locked output` (the last could infer that the outpoint exists but is unlocked).

Reply:
Makes sense, I'll make the errors more precise.